### ISSUE TITLE
Having issues with hmmlearn: only works with square input matrix

### ISSUE THREAD

**Issue by kmthayer** [hmmlearn/hmmlearn#509]
Hello,
I am trying to morph a Multinomial HMM from an example I found online to one for my research (link to original tutorial in header of sample code below). It has the number of datapoints as 5 and the number of observations also equal to 5. These are observed words in a conversation about dogs and cats. In my application, my observed "words" are the DNA bases A, G, C and T.  It is fine to have a fixed number of observations per sequence (the columns), but I will have many sequences, ie the input matrix grows an arbitrary number of rows.  So long as the input matrix (rawdata below) is square and larger than 1 (tutorial example was 5x5), it seems to work fine, but when I change this as I need to do for the application, I keep getting the error **ValueError Total Count for each sample should add up to the number of trials**.  I tracked this to hmm.py to this conditional 
if not (X.sum(axis=1) == self.n_trials).all():
To me it seems X.sum(axis=1) returns a list with the sum of the total number of times each DNA base was seen. Since they all have the same number of observations (columns) and each observation counts for 1, the entries in the matrix are all equal, which all makes sense. the self.n_trials seems to be an integer although from reading the doc I thought it would be a list having the number of columns repeated as many times as the number of rows I have.  It seems to represent the number of rows in the input matrix as a single integer. But I am puzzled as to why we would expect the entries across the columns to equal the number of rows; to me it seems the test one would want to do is whether the X.sum(axis=1) equals the number of columns, not the rows (number of promoters with DNA sequences).  Indeed, a square input would make the statement True as I saw. I feel that I am missing something critical here; that it should be possible to not have a square input matrix so that more DNA sequence data can be used to train the model.  
Something seems to be lacking on my understanding on how to add more data such that the input has more rows than columns, and still have the model  work.  Thanks so much for any advice you can give.

```python
# -*- coding: utf-8 -*-
"""
Created on Wed Mar 15 13:23:14 2023

@author: Kelly
"""

# -*- coding: utf-8 -*-
"""
pyHMM_capseq
Created on Mon Mar 13 08:13:36 2023
Original tutorial
https://hmmlearn.readthedocs.io/en/latest/auto_examples/plot_multinomial_hmm.html#sphx-glr-auto-examples-plot-multinomial-hmm-py
@author: Kelly
@author: Kelly
the goal here is to make a multinomial HMM to train on CAP sequences
"""
import numpy as np
import pandas as pd
from hmmlearn import hmm


def get_dfs_from_excel(infile):#,tabs):
    """
    sig: str,list[str]->NoneType
    creates global dataframe containing the excel data
    infile is the name of the excel input, 
    df is global variable the dataframe
    """
    print('Importing data from',infile,'to dataframes.')
    global df# a pandas dataframe
    #for tab in tabs:
       #list_dfs.append(pd.read_excel(infile,sheet_name=tab)) 
    df=pd.read_excel(infile,sheet_name='CAP')


def get_df_info(df):
    """sig:pandas dataframe-> noneType
    Get info on the dataframe df and print to screen"""
    print(df.head(10)) # first few lines
    print(df.shape)# tells size, does not use () for some reason...?
    print(df.dtypes)# discover types of data
    print(df.describe)# prints out stats. since we have str not much.
    #print(df['Name'][19],df['Sequence'][19]) # how to index stuff
    
def get_model_info(HMM):
    """MultinomialHMM object -> NoneType
    get list of all the variables and 
    read the doc
    """
    print(dir(model)) # list of all defined vars about it
    help(HMM) # read doc

# the input file, but I actually hard coded the seq as rawdata below :)    
infile = r'C:\Users\Kelly\anaconda3\Lib\site-packages\PyHMM\CAPseqRealExcel.xlsx'
# import data to train the HMM
get_dfs_from_excel(infile)
# get information about the dataframe using pandas tools
get_df_info(df)

# For this example, we will model the stages of a conversation,
# where each sentence is "generated" with an underlying topic, "cat" or "dog"
# For the CAP, we will model sequence of CAP binding site
# where each input sequence is generated with an underlying DNA 
# alphabet character AGCT
#states = ["cat", "dog"]
states = ['M1','M2']# a naive model with just 2 main states
id2topic = dict(zip(range(len(states)), states))# think this gives a number
# we are more likely to talk about cats first
start_probs = np.array([0.6, 0.4]) # start transition to each state.

# For each topic, the probability of saying certain words can be modeled by
# a distribution over vocabulary associated with the categories
# emissions letters
#vocabulary = ["tail", "fetch", "mouse", "food"]
vocabulary = ['A','G','C','T']

# if the topic is "cat", we are more likely to talk about "mouse"
# if the topic is "dog", we are more likely to talk about "fetch"
# I will make the priors be flat
emission_probs = np.array([[0.25, 0.25, 0.25, 0.25],
                           [0.25, 0.25, 0.25, 0.25]])

# Also assume it's more likely to stay in a state than transition to the other
# this is the transition matrix. Flat prob is fine to start for me.
trans_mat = np.array([[0.5, 0.5], [0.5, 0.5]])


# Pretend that every sentence we speak only has a total of 5 words,
# i.e. we independently utter a word from the vocabulary 5 times per sentence
# we observe the following bag of words (BoW) for 8 sentences:
# now it gets tricky. I may just hard doce some data or try to pull from 
# the dataframe
"""
observations = [["tail", "mouse", "mouse", "food", "mouse"],
        ["food", "mouse", "mouse", "food", "mouse"],
        ["tail", "mouse", "mouse", "tail", "mouse"],
        ["food", "mouse", "food", "food", "tail"],
        ["tail", "fetch", "mouse", "food", "tail"],
        ["tail", "fetch", "fetch", "food", "fetch"],
        ["fetch", "fetch", "fetch", "food", "tail"],
        ["food", "mouse", "food", "food", "tail"],
        ["tail", "mouse", "mouse", "tail", "mouse"],
        ["fetch", "fetch", "fetch", "fetch", "fetch"]]
"""
# so my task will be to convert from the string to the list of lists
# lets begin with a few hard coded ones.
"""
rawdata={'acs1':list('CTTTTGCGTGATCTGTCGCCCAAA'),
         'acs2':list('TTTGAAGGTCAGCAAC'),
         'aldA':list('AATTCTGTGAAGGGCTTCATAAAAACAA'),
         'aldB':list('TTGCCCGTAAATTCGTGATAGCTGTCGTAAAGCT'),
         'araBAD1':list('TGCCGTGATTATAGACACTTTTGTTA')}"""
# could the problem total cound for each sample should ad up to the number of trials
# YES. each of these seq must be the exact same length.
rawdata={'acs1':list('CGA'), #A
         #'acs2':list('TGT'), #A
         #'aldA':list('TGT'), #A
         'fak2':list('TGT'),
         'aldB':list('CGTGAA'), #A
         'fake':list('CGT')} #A
         #'fak2':list('GAGAG')} # having all A causes valueError SOMETIMES
         #'araBAD1':list('CGTGA')}#    
         #'fake':list('AAAAA')}

def raw2LofLdata(rawdata):
    """sig: dict{str:list[seqstr]-> list[list[str]]
    converts raw data as dict into a list of lists. This will have just
    the sequences as a list of lists of each alphabetic character"""
    acc=[]
    for bdsite in rawdata:
        acc.append(rawdata[bdsite])
    return acc

# convert data from rawdata dict into the list of lists LoLdata # list of list data
LoLdata=raw2LofLdata(rawdata) # observations in original tutorial

# Convert "sentences" to numbers:
# The DNA sequences are getting converted into numeric data

# this part just makes a dictionary to map the alphabet to numbers A=0, G=1 C=2 T=3
vocab2id = dict(zip(vocabulary, range(len(vocabulary))))

def sentence2counts(sentence):
    """sig: list[list[st]r]->list[int]
    the list of list contains single alphabetic characters of DNA seq
    returns the counts of each of the letters"""
    ans = []
    # dict.items converts entries into a list of tuples. then we match the 
    # indices to the tuple entries
    for word, idx in vocab2id.items():
        count = sentence.count(word) # take word part of tuple & count
        ans.append(count)
    return ans

# we next get a list of lists with counts for the AGCT letters
X = []
for seq in LoLdata:
    row = sentence2counts(seq)
    X.append(row)

# now for some reason we feel the need to convert to numpy array
data = np.array(X, dtype=int)
# OKAY... so this is in fact the sum of the word frequencies, ie how many As
# not exactly what I want. I think the training set I want is actually 
# quite more simple, just the raw sequence of characters
# but let us go with this for the moment to check it will execute.

# suppose I want to skip data enhancement altogether
"""
# data enhancement
# pretend this is repeated, so we have more data to learn from:
lengths = [len(X)]*5 # how long it will be after it is tiled=len(sequences)
sequences = np.tile(data, (5,1)) # tiling data 5x1 sort of like MATLAB
"""
# circumvent data enhancement
sequences=data
lengths=[len(X)]


# Set up model:
model = hmm.MultinomialHMM(n_components=len(states),
        n_trials=len(LoLdata[0]), # how many char long is seq? EX: GAATTCGCG=9
        n_iter=50, # number of training iterations to perform
        init_params='') # which STE used for init phase. or just start w regular probs

model.n_features = len(vocabulary) # the vocab or alphabet are the features
model.startprob_ = start_probs # probs between start state and each state
model.transmat_ = trans_mat # transition probs amongst the states
model.emissionprob_ = emission_probs # prob to emit alphabet vocab by state


# Do the model fitting
#model.fit(data,lengths)
#model.fit(sequences, lengths) # got error here. Make all seq same length
model.fit(sequences)
l

```
